File: BackEnd/TelegramBot.Bot/ChannelPostHandler.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Telegram.Bot;
using Telegram.Bot.Types;
using Telegram.Bot.Types.ReplyMarkups;
using TelegramBot.Application.Interfaces;
using TelegramBot.Application.Services.PostService;
using TelegramBot.Core.Entities;

namespace TelegramBot.Bot
{
    public class ChannelPostHandler
    {
        IPostService postService;
        BotSettingsEntity botSettings;
        ILoggerService logger;
        private readonly ITelegramBotClient botClient;
        private readonly Update update;
        private readonly CancellationToken cancellationToken;

        public ChannelPostHandler(IPostService postService,
                                    ITelegramBotClient botClient, Update update, CancellationToken cancellationToken, BotSettingsEntity botSettings, ILoggerService logger)
        {
            this.postService = postService;
            this.botClient = botClient;
            this.update = update;
            this.cancellationToken = cancellationToken;
            this.botSettings = botSettings;
            this.logger = logger;
        }

        bool CheckMentionChannelTag(List<string> Entities, string channelId)
        {
            foreach (var entity in Entities)
            {
                if (entity.Equals($"@{channelId.ToLower()}", StringComparison.CurrentCultureIgnoreCase))
                {
                    return true;
                }
            }

            return false;
        }

        public async void OnChannelPost(Message post, List<long> ignoreList)
        {
            var chatId = post.Chat.Id;
            var messageId = post.MessageId;

            PostInfoEntity postInfoEntity = postService.SaveNewPost(new PostInfoEntity()
            {
                TelegramId_PostId = (int)post.ForwardFromMessageId,
                TelegramId_ChatId = post.ForwardFromChat.Id,
            });

            List<string> entities = new();

            if (post.ForwardFromChat is { } forwardedChat)
            {
                if (ignoreList.Contains(forwardedChat.Id))
                {
                    return;
                }
            }

            if (post.CaptionEntityValues != null)
            {
                entities = post.CaptionEntityValues.ToList();
            }
            else if (post.EntityValues != null)
            {
                entities = post.EntityValues.ToList();
            }

            var inlineKeyboard = Utility.getInlineKeyboard(postInfoEntity.Id);

            var mentioned = CheckMentionChannelTag(entities, botSettings.MainChannelUsername);
            Console.WriteLine($"Channel Mentioned : {mentioned}");

            if (mentioned)
            {
                Message sentMessage = new Message();

                if (botSettings.AlternativeMainGroupID != 0)
                {
                    sentMessage = await botClient.ForwardMessageAsync(
                    chatId: botSettings.AlternativeMainGroupID,
                    messageThreadId: botSettings.AlternativeMainGroupTopicID,
                    fromChatId: chatId,
                    messageId: messageId,
                    cancellationToken: cancellationToken);
                }
                else
                {
                    sentMessage = await botClient.ForwardMessageAsync(
                        chatId: botSettings.MainGroupID,
                        fromChatId: chatId,
                        messageId: messageId,
                        cancellationToken: cancellationToken);
                }

                var logMessage = $"Forward to group: {messageId} , {chatId} , postid: {postInfoEntity.Id}";
                logger.LogInfo(logMessage);

                Message markupMessage = await botClient.SendTextMessageAsync(
                    chatId: sentMessage.Chat.Id,
                    text: "وضعیت پست:",
                    replyMarkup: inlineKeyboard,
                    replyToMessageId: sentMessage.MessageId,
                    cancellationToken: cancellationToken);

                PostInfoTelegramModel postInfoTelegramModel = new()
                {
                    TelegramId_ChatId = markupMessage.Chat.Id,
                    TelegramId_PostId = markupMessage.MessageId,
                    Repliyed_TelegramId_ChatId = sentMessage.Chat.Id,
                    Repliyed_TelegramId_PostId = sentMessage.MessageId
                };

                postService.ForwardToGroupPost(postInfoEntity, postInfoTelegramModel);

                var oldPostInfos = postService.GetEventsByTelegramId(postInfoEntity.TelegramId_ChatId, postInfoEntity.TelegramId_PostId).ToList();
                foreach (var oldDatas in oldPostInfos)
                {
                    if (oldDatas.Id != postInfoEntity.Id)
                        try
                        {
                            var postEvents = postService.GetRemovableEvents(oldDatas.Id);
                            foreach (var postEvent in postEvents)
                            {
                                await botClient.DeleteMessageAsync(postEvent.TelegramId_ChatId, postEvent.TelegramId_PostId, cancellationToken);
                                await botClient.DeleteMessageAsync(postEvent.Repliyed_TelegramId_ChatId, postEvent.Repliyed_TelegramId_PostId, cancellationToken);
                            }
                            postService.RemoveOldPostStatus(oldDatas);
                        }
                        catch (Exception ex)
                        {
                            Console.WriteLine(ex.Message);
                            return;
                        }
                }
            }
        }

        public void OnEditedChannelPost()
        {

        }
    }
}

File: Bot/Commands/ChannelPostHandler.py

from telethon import TelegramClient, functions
from telethon.types import UpdateNewChannelMessage

from apihelper.BotSettingsServiceApi import BotSettingsServiceApi
from apihelper.PostServiceApi import PostServiceApi
from Commands.MessageStatus import MessagesStatus
from utility.inline_replies import get_inline_keyboard


class ChannelPostHandler:
    def __init__(self, client: TelegramClient, 
                 bot_settings_service : BotSettingsServiceApi, 
                 post_service : PostServiceApi):
        
        self.bot_client = client
        self.bot_settings_service = bot_settings_service
        self.post_service = post_service
    
    def check_mention_channel_tag(self, text, channel_id):
        if f"@{channel_id.lower()}" in text.lower():
            return True
        return False

    async def on_channel_post(self , post, ignore_list):
        
        botsetting = self.bot_settings_service.get_bot_settings()
        
        org_message_ids = [x.forward.channel_post for x in post]
        
        brdg_message_ids = [x.id for x in post]

        mainpost = post[0]
        
        post_info_entity = self.post_service.create_new_post({
            "TelegramId_PostId":mainpost.forward.channel_post,
            "TelegramId_ChatId":mainpost.forward.chat_id,
            "Telegram_mediagroup_postids":org_message_ids,
            "Media_groupId":0,
        })

        if mainpost.forward.chat_id in ignore_list:
            return

        inline_keyboard = get_inline_keyboard(post_info_entity["id"])

        mentioned = self.check_mention_channel_tag(mainpost.message, botsetting["mainChannelUsername"])
        print(f"Channel Mentioned: {mentioned}")

        if mentioned:
            try:
                
                chatId = botsetting["alternativeMainGroupID"] if botsetting["alternativeMainGroupID"] != 0 else botsetting["mainGroupID"]
                
                fwd_msg_events = await self.bot_client(functions.messages.ForwardMessagesRequest(
                    from_peer=mainpost.chat_id,
                    id=brdg_message_ids,
                    to_peer=chatId,
                    top_msg_id=botsetting["alternativeMainGroupTopicID"],
                ))
                
                for updt in fwd_msg_events.updates:
                    if type(updt) == UpdateNewChannelMessage and updt.message != '':
                        sent_message = updt.message
                        break
            

                # log_message = f"Forward to group: {message_id}, {chat_id}, postid: {post_info_entity.id}"
                # logger.log_info(log_message)

                markup_message = await self.bot_client.send_message(
                    entity=sent_message.chat_id,
                    message=MessagesStatus.pending_message(mainpost.forward.date),
                    reply_to=sent_message.id,
                    buttons=inline_keyboard
                )

                post_info_telegram_model = {
                    "TelegramId_ChatId":markup_message.chat_id,
                    "TelegramId_PostId":markup_message.id,
                    "Repliyed_TelegramId_ChatId":sent_message.chat_id,
                    "Repliyed_TelegramId_PostId":sent_message.id
                }

                self.post_service.forward_to_group_post(post_info_entity, post_info_telegram_model)

                old_post_infos = self.post_service.get_events_by_telegram_id(post_info_entity["telegramId_ChatId"], post_info_entity["telegramId_PostId"])
                for old_data in old_post_infos:
                    if old_data["id"] != post_info_entity["id"]:
                        try:
                            post_events = self.post_service.get_removable_events(old_data["id"])
                            for post_event in post_events:
                                await self.bot_client.delete_messages(post_event["telegramId_ChatId"], [post_event["telegramId_PostId"], post_event["repliyed_TelegramId_PostId"]])
                            self.post_service.remove_old_post_status(old_data)
                        except Exception as ex:
                            print(ex)
                            return
            except Exception as ex:
                print(ex)
                return